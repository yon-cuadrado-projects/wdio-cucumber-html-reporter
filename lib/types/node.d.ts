import { Browser } from 'webdriverio';
import { cjson_metadata } from '../models';
export { };

declare global {
    namespace NodeJS {
        export interface EnvironmentVariables {
            wdioCucumberHtmlReporter: string;
            attachment: string;
        }

        export interface Global {
            browser: Browser<'sync'>;
        }
    }

    namespace WebDriver {
        export interface W3CCapabilities {
            cjson_metadata: cjson_metadata;
        }
    }
}
