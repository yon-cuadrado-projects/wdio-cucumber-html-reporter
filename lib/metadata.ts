import { DesiredCapabilitiesExtended, RunnerStatsExtended, W3CCapabilitiesExtended, WebdriverIOExtended } from './types/wdio';
import { Browser } from 'webdriverio';
import { Models } from 'cucumber-html-report-generator';
import { NOT_KNOWN } from './constants';
import WebDriver from 'webdriver';
import { cjson_metadata } from './models';

export class Metadata {
    public determineMetadata ( data: RunnerStatsExtended ): Models.Metadata[] {
        const featureMetadata = <Models.Metadata[]>[];
        const currentCapabilities = data.capabilities as W3CCapabilitiesExtended;
        let w3cCaps = ( < W3CCapabilitiesExtended | DesiredCapabilitiesExtended>( data.config.capabilities ) )?.['cjson:metadata'];
        const currentConfigCapabilities = data.capabilities as DesiredCapabilitiesExtended;
        const metadataFromWdioConfiguration: cjson_metadata = currentConfigCapabilities?.['cjson:metadata'];

        if( typeof browser !== 'undefined' ){
            w3cCaps = w3cCaps ?? ( ( browser as Browser<'async'> ).options as WebdriverIOExtended )?.requestedCapabilities?.['cjson:metadata'];
            const optsCaps = browser.options.capabilities;
            const browserMetadata = metadataFromWdioConfiguration ?? ( optsCaps as DesiredCapabilitiesExtended )?.['cjson:metadata'] ?? w3cCaps;
            const instanceBrowserData = this.determineBrowserData( currentCapabilities, currentConfigCapabilities, browserMetadata );
            featureMetadata.push( instanceBrowserData );
        } else{
            const instanceAppData = this.determineAppData( currentConfigCapabilities, metadataFromWdioConfiguration );
            featureMetadata.push( instanceAppData );
        }

        const device = this.determineDeviceName( metadataFromWdioConfiguration, currentConfigCapabilities );
        const platformName = this.determinePlatformName( metadataFromWdioConfiguration, currentCapabilities );
        const icon = this.getIcon( platformName );

        const currentPlatform = {
            name: platformName,
            version: this.determinePlatformVersion( metadataFromWdioConfiguration ),
        };

        featureMetadata.push( <Models.Metadata>{ name: 'device', value: device } );
        featureMetadata.push( <Models.Metadata>{ name: 'platform', value: `${currentPlatform.name} ${currentPlatform.version}`, icon } );

        return featureMetadata;
    }

    public determineDeviceName ( metadata: cjson_metadata, currentConfigCapabilities: WebDriver.DesiredCapabilities ): string {
        return ( metadata?.device || currentConfigCapabilities.deviceName || `Device name ${NOT_KNOWN}` );
    }

    public determinePlatformName ( metadata: cjson_metadata, currentCapabilities: WebDriver.DesiredCapabilities ): string {
        const currentPlatformName = currentCapabilities.platformName
            ? currentCapabilities.platformName.includes( 'mac' )
                ? 'osx'
                : currentCapabilities.platformName.includes( 'windows' )
                    ? 'windows'
                    : currentCapabilities.platformName
            : `Platform name ${NOT_KNOWN}`;


        return metadata?.platform?.name ?? currentPlatformName;
    }

    public determinePlatformVersion ( metadata: cjson_metadata ): string {
        return metadata?.platform?.version ?? `Version ${NOT_KNOWN}`;
    }

    public determineAppData ( currentConfigCapabilities: DesiredCapabilitiesExtended, metadata: cjson_metadata ): Models.Metadata {
        const metaAppName: string = metadata?.app?.name ?? 'No metadata.app.name available';
        const metaAppVersion: string = metadata?.app?.version ?? 'No metadata.app.version available';
        const appPath = currentConfigCapabilities.app || currentConfigCapabilities.testobject_app_id || metaAppName;
        const appName = appPath.substring( appPath.replace( '\\', '/' ).lastIndexOf( '/' ) ).replace( '/', '' );

        return {
            name: 'app', value: `${appName} ${metaAppVersion}`
        };
    }

    public determineBrowserData ( currentCapabilities: WebDriver.DesiredCapabilities, currentConfigCapabilities: WebDriver.DesiredCapabilities, metadata: cjson_metadata ): Models.Metadata {
        const browserName = currentCapabilities.browserName
            || currentConfigCapabilities.browserName
            || ( metadata?.browser?.name ?? 'No metadata.browser.name available' );
        const browserVersion = currentCapabilities.version
            || currentCapabilities.browserVersion
            || currentConfigCapabilities.browserVersion
            || ( metadata?.browser?.version ?? 'No metadata.browser.version available' );
        const icon = this.getIcon( browserName );

        return{
            name: 'browser', value: `${browserName} ${browserVersion}`, icon
        };
    }

    /* istanbul ignore next */
    private getIcon( value: string ): string{
        let icon = '';
        switch( value.trim() ){
        case 'internet explorer':
            icon = 'fab fa-internet-explorer';
            break;
        case 'safari':
            icon = 'fab fa-safari';
            break;
        case 'edge':
            icon = 'fab fa-edge';
            break;
        case 'chrome':
            icon = 'fab fa-chrome';
            break;
        case 'firefox':
            icon = 'fab fa-firefox';
            break;
        case 'windows':
            icon = 'fab fa-windows';
            break;
        case 'osx':
        case 'ios':
            icon = 'fab fa-apple';
            break;
        case 'linux':
            icon = 'fab fa-linux';
            break;
        case 'ubuntu':
            icon = 'fab fa-ubuntu';
            break;
        case 'android':
            icon = 'fab fa-android';
            break;
        }

        return icon;
    }
}
