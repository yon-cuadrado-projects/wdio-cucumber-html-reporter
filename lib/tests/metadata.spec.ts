import { SMALL_RUNNER_STATS, SMALL_RUNNER_STATS_PLATFORM_MAC, SMALL_RUNNER_STATS_PLATFORM_OSX, SMALL_RUNNER_STATS_PLATFORM_WINDOWS, SMALL_RUNNER_STATS_WITH_CJSON } from './__mocks__/mocks';
import { Browser } from 'webdriverio';
import { Metadata } from '../metadata';
import { WebdriverIOExtended } from '../types/wdio';

describe( 'metadata', () => {
    let metadataClassObject: Metadata;
    beforeAll( () => {
        metadataClassObject = new Metadata();
    } );

    describe( 'determineMetadata', () => {
        beforeEach( () => {
            delete global.browser;
            global.browser = {
                options: {
                    requestedCapabilities: {
                        'cjson:metadata': {
                            app: {
                                name: 'mock-appName',
                                version: 'mock-appVersion',
                            },
                        },
                        w3cCaps: {
                            alwaysMatch: {
                                foo: true,
                            }
                        }
                    },
                } as WebdriverIOExtended,
            } as Browser<'async'>;

        } );

        afterEach( () => {
            delete global.browser;
        } );

        it( 'should return no metadata when there is none', () => {
            global.browser = {
                options: {
                } as WebdriverIOExtended,
            } as Browser<'async'>;
            expect( metadataClassObject.determineMetadata( SMALL_RUNNER_STATS_WITH_CJSON ) ).toMatchSnapshot();
        } );

        it( 'should return app metadata with cjson:metadata', () => {
            const result = metadataClassObject.determineMetadata( SMALL_RUNNER_STATS );
            expect( result ).toMatchSnapshot();
        } );

        it( 'should return metadata with platform mac', () => {
            const result = metadataClassObject.determineMetadata( SMALL_RUNNER_STATS_PLATFORM_MAC );
            expect( result ).toMatchSnapshot();
        } );

        it( 'should return metadata with platform osx', () => {
            const result = metadataClassObject.determineMetadata( SMALL_RUNNER_STATS_PLATFORM_OSX );
            expect( result ).toMatchSnapshot();
        } );

        it( 'should return metadata with platfrom windows', () => {
            const result = metadataClassObject.determineMetadata( SMALL_RUNNER_STATS_PLATFORM_WINDOWS );
            expect( result ).toMatchSnapshot();
        } );
    } );
} );
