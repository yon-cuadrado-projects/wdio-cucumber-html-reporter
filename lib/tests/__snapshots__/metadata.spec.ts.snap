// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata determineMetadata should return app metadata with cjson:metadata 1`] = `
Array [
  Object {
    "icon": "",
    "name": "browser",
    "value": "No metadata.browser.name available No metadata.browser.version available",
  },
  Object {
    "name": "device",
    "value": "Device name not known",
  },
  Object {
    "icon": "",
    "name": "platform",
    "value": "Platform name not known Version not known",
  },
]
`;

exports[`metadata determineMetadata should return metadata with platform mac 1`] = `
Array [
  Object {
    "icon": "",
    "name": "browser",
    "value": "No metadata.browser.name available No metadata.browser.version available",
  },
  Object {
    "name": "device",
    "value": "Device name not known",
  },
  Object {
    "icon": "fab fa-apple",
    "name": "platform",
    "value": "osx Version not known",
  },
]
`;

exports[`metadata determineMetadata should return metadata with platform osx 1`] = `
Array [
  Object {
    "icon": "",
    "name": "browser",
    "value": "No metadata.browser.name available No metadata.browser.version available",
  },
  Object {
    "name": "device",
    "value": "Device name not known",
  },
  Object {
    "icon": "fab fa-apple",
    "name": "platform",
    "value": "osx Version not known",
  },
]
`;

exports[`metadata determineMetadata should return metadata with platfrom windows 1`] = `
Array [
  Object {
    "icon": "",
    "name": "browser",
    "value": "No metadata.browser.name available No metadata.browser.version available",
  },
  Object {
    "name": "device",
    "value": "Device name not known",
  },
  Object {
    "icon": "fab fa-windows",
    "name": "platform",
    "value": "windows Version not known",
  },
]
`;

exports[`metadata determineMetadata should return no metadata when there is none 1`] = `
Array [
  Object {
    "icon": "",
    "name": "browser",
    "value": "No metadata.browser.name available No metadata.browser.version available",
  },
  Object {
    "name": "device",
    "value": "Device name not known",
  },
  Object {
    "icon": "",
    "name": "platform",
    "value": "Platform name not known Version not known",
  },
]
`;
